tokio/src/task/mod.rs

```diff
@@ -269,42 +264,65 @@ cfg_rt_util! {
     pub use task_local::LocalKey;
 }
 
+cfg_rt_threaded! {
+    mod stack;
+    pub(crate) use self::stack::TransferStack;
+}
+
 cfg_rt_core! {
+    /*
     /// Unit tests
     #[cfg(test)]
     mod tests;
+    */
 
     use std::future::Future;
     use std::marker::PhantomData;
     use std::ptr::NonNull;
     use std::{fmt, mem};
 
     /// An owned handle to the task, tracked by ref count
+    #[repr(transparent)]
     pub(crate) struct Task<S: 'static> {
         raw: RawTask,
         _p: PhantomData<S>,
     }
 
-    unsafe impl<S: ScheduleSendOnly + 'static> Send for Task<S> {}
+    unsafe impl<S: ScheduleSendOnly> Send for Task<S> {}
+    unsafe impl<S: ScheduleSendOnly> Sync for Task<S> {}
+
+    /// A task was notified
+    #[repr(transparent)]
+    pub(crate) struct Notified<S: 'static>(Task<S>);
+
+    unsafe impl<S: Schedule> Send for Notified<S> {}
+    unsafe impl<S: Schedule> Sync for Notified<S> {}
 
     /// Task result sent back
     pub(crate) type Result<T> = std::result::Result<T, JoinError>;
 
-    pub(crate) trait Schedule: Sized + 'static {
+    pub(crate) trait Schedule: Sync + Sized + 'static {
         /// Bind a task to the executor.
         ///
         /// Guaranteed to be called from the thread that called `poll` on the task.
-        fn bind(&self, task: &Task<Self>);
+        fn bind(task: &Task<Self>) -> Self;
```

Comment:
This is a really weird signature now. Some more docs for why it is what it is would be helpful.

Reply:
What is weird about it?

Reply:
Added:

```
    /// Bind a task to the executor.
    ///
    /// Guaranteed to be called from the thread that called `poll` on the task.
    /// The returned `Schedule` instance is associated with the task and is used
    /// as `&self` in the other methods on this trait.
```

Reply:
It's a signature that does not take a `&self`, and needs to return a `Self`. The name `bind` does not seem to accurately reflect what it does, at least not trivially. It looks more like it "extracts" the scheduler from a task or something?